### PR TITLE
fix: use xhr to check container type of sidx and blacklist if anything but mp4

### PR DESCRIPTION
**Pull request by brandonocasey**
A pull request to work-around webm content that we don't currently support. This pull request checks the container format of the sidx segment and blacklists the playlist if it is not mp4.
 
Depends on https://github.com/videojs/vhs-utils/pull/12

### CODE REVIEW

File: src/util/codecs.js [hunk 1]
```diff
@@ -121,3 +122,52 @@ export const isLikelyWebmData = (bytes) => {
     (bytes[2] & 0xFF) === 0xDF &&
     (bytes[3] & 0xFF) === 0xA3;
 };
+
+export const isLikelyMp4Data = (bytes) => {
```
Comment (brandonocasey) [inline src/util/codecs.js:126]: Long ago in tba this function was called `isLikelyFmp4Data` but I had to change it to the current iteration because it also matched regular mp4 data. Now I am bringing it back as I think it fits our use case here.

**Comment by brandonocasey** [comment]
This pull request needs:
1. Test fixes
2. cleanup the sidx request portion of the code as its very hard to follow

A Future pull request could:
Use this new function to determine the container type without downloading the whole segment in mediaSegmentRequest. That way we can blacklist after only downloading 8 bytes plus a playlist rather than an entire segment, or even two segments.

File: src/segment-loader.js [hunk 2]
```diff
@@ -992,7 +992,7 @@ export default class SegmentLoader extends videojs.EventTarget {
       this.sourceUpdater_.removeAudio(start, end, removeFinished);
     }
 
-    if (this.loaderType_ === 'main' && this.startingMedia_.hasVideo) {
+    if (this.loaderType_ === 'main' && this.startingMedia_ && this.startingMedia_.hasVideo) {
```
Comment (brandonocasey) [inline src/segment-loader.js:995]: protect against undefined `this.startingMedia_` as that is the state it starts in.

File: src/util/codecs.js [hunk 3]
```diff
@@ -105,19 +105,3 @@ export const isLikelyFmp4Data = (bytes) => {
...
- * Check to see if the bytes are part of a WebM media file.
- *
- * @param {Uint8Array} bytes
- *        The starting bytes of the file, generally from the map, or media initialization
- *        section, of the file.
- * @return {Boolean}
- *         Whether the bytes likely come from a WebM media file.
- * @see https://en.wikipedia.org/wiki/List_of_file_signatures
- */
-export const isLikelyWebmData = (bytes) => {
```
Comment (brandonocasey) [inline src/util/codecs.js:118]: moved to vhs-utils.

File: test/dash-playlist-loader.test.js [hunk 4]
```diff
@@ -19,6 +19,7 @@ import {
 import '../src/videojs-http-streaming';
 import testDataManifests from 'create-test-data!manifests';
 import { sidx as sidxResponse } from 'create-test-data!segments';
+import {mp4VideoInit as mp4VideoInitSegment} from 'create-test-data!segments';
```
Comment (brandonocasey) [inline test/dash-playlist-loader.test.js:22]: Now we have to supply a legit init segment for sidx requests or else we blacklist the playlist. Had to update tests to do that.

File: src/dash-playlist-loader.js [hunk 5]
```diff
@@ -393,7 +418,8 @@ export default class DashPlaylistLoader extends EventTarget {
       sidxInfo: playlist.sidx
     };
 
-    this.request = requestSidx_(
+    this.request = requestSidx_.call(
+      this, startingState,
```
Comment (gkatsev) [inline src/dash-playlist-loader.js:422]:
```suggestion
      this,
      startingState,
```

File: src/dash-playlist-loader.js [hunk 6]
```diff
@@ -306,7 +345,7 @@ export default class DashPlaylistLoader extends EventTarget {
         return doneFn(master, null);
       }
 
-      const bytes = new Uint8Array(request.response);
+      const bytes = toUint8(request.response);
```
Comment (brandonocasey) [inline src/dash-playlist-loader.js:348]: Since we potentially re-use bytes and they can already be a uin8 at this point, we use the helper from vhs-utils which only creates a new uint8 if needed.

File: src/xhr.js [hunk 7]
```diff
@@ -16,6 +16,45 @@ const {
   mergeOptions
 } = videojs;
 
+const callbackWrapper = function(request, error, response, callback) {
```
Comment (brandonocasey) [inline src/xhr.js:19]: This code is unchanged, I just moved it into its own function so that container-request can use it.

File: test/dash-playlist-loader.test.js [hunk 8]
```diff
@@ -471,7 +472,7 @@ QUnit.test('requestSidx_: creates an XHR request for a sidx range', function(ass
   const sidxInfo = {
     resolvedUri: 'sidx.mp4',
     byterange: {
-      offset: 0,
+      offset: 10,
```
Comment (brandonocasey) [inline test/dash-playlist-loader.test.js:475]: We don't want to re-use bytes in this test (from the container request) so I set the offset so that it doesn't happen.

File: test/test-helpers.js [hunk 9]
```diff
@@ -372,31 +373,29 @@ export const standardXHRResponse = function(request, data) {
...
-    response = data.buffer;
-  } else if (!data) {
-    // A placeholder is used for some old tests. This may be a good target to clean up in
-    // the future.
-    response = new Uint8Array(1024).buffer;
-  } else {
-    response = data;
+  if (isTypedArray(response)) {
+    // a string for container requests or a buffer for non-container requests
+    response = isContainerRequest ? bytesToString(response) : response.buffer;
```
Comment (brandonocasey) [inline test/test-helpers.js:398]: container requests are `byte` strings, so that we can work-around the fact that xhr does not add to `request.response` until the request is done, it only adds to `request.responseText`, which throws an error when accesed if the `responseType` is `arraybuffer` like normal.

File: test/util/codecs.test.js [hunk 10]
```diff
@@ -379,30 +374,3 @@ QUnit.test('parses codecs regardless of codec order', function(assert) {
     'parses audio first'
   );
 });
-
-QUnit.module('isLikelyWebmData');
```
Comment (brandonocasey) [inline test/util/codecs.test.js:383]: All this moved to vhs-utils

File: src/dash-playlist-loader.js [hunk 11]
```diff
@@ -204,7 +206,44 @@ export const requestSidx_ = (sidxRange, playlist, xhr, options, finishProcessing
...
+      return finishProcessingFn(err, request);
+    }
+
+    if (!container || container !== 'mp4') {
+      this.error = {
+        status: request.status,
+        message: `Unsupported ${container || 'unknown'} container type for sidx segment at URL: ${request.uri}`,
+        response: request.response,
+        playlist,
+        internal: true,
```
Comment (brandonocasey) [inline src/dash-playlist-loader.js:220]: Are we ok with this being an internal error? Seems like a feature we lack rather than an actual error to me, but if we want to warn in the console that could be fine.

File: src/media-segment-request.js [hunk 12]
```diff
@@ -168,11 +168,14 @@ const handleInitSegmentResponse =
...
-  if (isLikelyWebmData(segment.map.bytes)) {
+  const type = detectContainerForBytes(segment.map.bytes);
+
+  if (type !== 'mp4') {
     return finishProcessingFn({
       status: request.status,
-      message: 'Found unsupported WebM initialization segment at URL: ' + request.uri,
+      message: `Found unsupported ${type || 'unknown'} container for initialization segment at URL: ${request.uri}`,
       code: REQUEST_ERRORS.FAILURE,
+      internal: true,
```
Comment (brandonocasey) [inline src/media-segment-request.js:178]: Are we ok with this being an internal error? Seems like a feature we lack rather than an actual error to me, but if we want to warn in the console that could be fine.

File: src/dash-playlist-loader.js (hunk 11 above)
Comment (gkatsev) [reply src/dash-playlist-loader.js:220]: I think we can just leave it as is for now, we probably want to rethink how errors are handled to be able to better allow for people to handle things as they want.

File: src/media-segment-request.js [hunk 13]
```diff
@@ -168,11 +168,14 @@ const handleInitSegmentResponse =
 
   segment.map.bytes = new Uint8Array(request.response);
 
-  if (isLikelyWebmData(segment.map.bytes)) {
+  const type = detectContainerForBytes(segment.map.bytes);
+
+  if (type !== 'mp4') {
```
Comment (gesinger) [inline src/media-segment-request.js:173]: A transport stream can still have a media initialization section, thus, an EXT-X-MAP tag: https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-3.2

File: src/segment-loader.js [hunk 14]
```diff
@@ -2317,8 +2317,8 @@ export default class SegmentLoader extends videojs.EventTarget {
     // Although transmuxing is done, appends may not yet be finished. Throw a marker
     // on each queue this loader is responsible for to ensure that the appends are
     // complete.
-    const waitForVideo = this.loaderType_ === 'main' && this.startingMedia_.hasVideo;
-    const waitForAudio = !this.audioDisabled_ && this.startingMedia_.hasAudio;
+    const waitForVideo = this.loaderType_ === 'main' && this.startingMedia_ && this.startingMedia_.hasVideo;
```
Comment (gesinger) [inline src/segment-loader.js:2320]: Is there a case where we're trying to perform this operation (and the check in updateTimingInfoEnd) without having startingMedia?

File: src/dash-playlist-loader.js [hunk 15]
```diff
@@ -204,7 +206,44 @@ export const requestSidx_ = (sidxRange, playlist, xhr, options, finishProcessing
...
   });
 
-  return xhr(sidxRequestOptions, finishProcessingFn);
+  return containerRequest(sidxInfo.uri, xhr, (err, request, container, bytes) => {
+    if (err) {
+      return finishProcessingFn(err, request);
+    }
+
+    if (!container || container !== 'mp4') {
+      this.error = {
```
Comment (gesinger) [inline src/dash-playlist-loader.js:215]: Because this is a standalone function, even though it can be `call`ed with a context, I think it would be preferable to pass in an error callback handler or use some other system than to use `this`.

File: src/util/container-request.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    request.abort();
+    const type = detectContainerForBytes(bytes);
+
+    cb(null, request, type, bytes);
+  };
+
+  const options = {
+    uri,
+    beforeSend(request) {
+      request.overrideMimeType('text/plain; charset=x-user-defined');
```
Comment (gesinger) [inline src/util/container-request.js:41]: Would be worth adding the comment used in the partial downloads elsewhere here to clarify why we're doing the override.

File: src/util/container-request.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    if (!id3Offset) {
+      id3Offset = getId3Offset(bytes);
+    }
+
+    // not enough data to determine type yet
+    if (bytes.length < id3Offset + 2) {
+      return;
+    }
+    request.abort();
+    const type = detectContainerForBytes(bytes);
```
Comment (gesinger) [inline src/util/container-request.js:33]: If there's no id3 offset, and bytes.length is 3, would that be enough bytes to determine the container in all cases?

File: src/util/container-request.js [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+  let id3Offset;
+
+  const progressListener = function(error, request) {
+    if (error) {
+      return cb(error, request, '', bytes);
+    }
+    const currentLength = bytes && bytes.length || 0;
+    const contentPart = request.responseText.substring(currentLength, request.responseText.length);
+    const newBytes = toUint8(stringToBytes(contentPart, true));
+    const temp = new Uint8Array(currentLength + newBytes.length);
```
Comment (gesinger) [inline src/util/container-request.js:16]: Might be good to use a more specific name, or specify why a temp var is being used.

File: test/test-helpers.js [hunk 19]
```diff
@@ -372,31 +373,29 @@ export const standardXHRResponse = function(request, data) {
     manifestName = request.url;
   }
 
+  const isContainerRequest = request.mimeTypeOverride === 'text/plain; charset=x-user-defined';
```
Comment (gesinger) [inline test/test-helpers.js:376]: Since this could be for partial data downloads in any case, we may want to rename to something like `isPartialRequest`

File: src/media-segment-request.js (hunk 13 above)
Comment (brandonocasey) [reply src/media-segment-request.js:173]: Won't that mean that the map is mp4 though?

File: src/segment-loader.js (hunk 14 above)
Comment (brandonocasey) [reply src/segment-loader.js:2320]: I only really see it when we are going to player error in some way, or when we try to load something that we shouldn't be, like loading a webm in the muxer.

File: src/util/container-request.js (hunk 17 above)
Comment (brandonocasey) [reply src/util/container-request.js:33]: Only mp3 and aac should have id3 offset and they only need two bytes after the id3 data to determine the type

File: src/media-segment-request.js (hunk 13 above)
Comment (gesinger) [reply src/media-segment-request.js:173]: The map would be a transport stream too, just with PAT and PMT

File: src/segment-loader.js (hunk 14 above)
Comment (gesinger) [reply src/segment-loader.js:2320]: Would it be better to fail quickly with an exception there? Since it can otherwise lead to undefined behavior, and will help to surface issues where we might be trying to do something that we shouldn't be doing.

File: src/media-segment-request.js (hunk 13 above)
Comment (brandonocasey) [reply src/media-segment-request.js:173]: The problem with that though is that we won't be able to parse it as we assume that the map is mp4 below. I will add a TODO comment specifying that we should also handle ts segment maps.

File: src/util/container-request.js (hunk 18 above)
Comment (brandonocasey) [reply src/util/container-request.js:22]: added a comment

File: src/segment-loader.js (hunk 14 above)
Comment (brandonocasey) [reply src/segment-loader.js]: OK I triggered an error for this case, probably a good idea to blacklist such a playlist and try other in this case too.

File: src/segment-loader.js [hunk 20]
```diff
@@ -2314,11 +2314,19 @@ export default class SegmentLoader extends videojs.EventTarget {
   }
 
   waitForAppendsToComplete_(segmentInfo) {
+    if (!this.startingMedia_) {
+      this.error({
+        message: 'No starting media returned, we dont support this media format.',
```
Comment (gesinger) [inline src/segment-loader.js:2319]:
```suggestion
        message: 'No starting media returned, likely due to an unsupported media format.',
```

File: src/dash-playlist-loader.js [hunk 21]
```diff
@@ -189,7 +191,7 @@ export const filterChangedSidxMappings = (masterXml, srcUrl, clientOffset, oldSi
 };
 
 // exported for testing
-export const requestSidx_ = (sidxRange, playlist, xhr, options, finishProcessingFn) => {
+export const requestSidx_ = function(loader, sidxRange, playlist, xhr, options, finishProcessingFn) {
```
Comment (gesinger) [inline src/dash-playlist-loader.js:194]:
```suggestion
export const requestSidx_ = (loader, sidxRange, playlist, xhr, options, finishProcessingFn) => {
```

File: src/media-segment-request.js [hunk 22]
```diff
@@ -168,11 +168,16 @@ const handleInitSegmentResponse =
 
   segment.map.bytes = new Uint8Array(request.response);
 
-  if (isLikelyWebmData(segment.map.bytes)) {
+  const type = detectContainerForBytes(segment.map.bytes);
+
+  // TODO: We should also handle ts init segments here, but we
+  // only know how to parse ts maps at the moment
```
Comment (gesinger) [inline src/media-segment-request.js:174]:
```suggestion
  // only know how to parse mp4 maps at the moment
```

File: src/segment-loader.js [hunk 23]
```diff
@@ -2440,7 +2448,7 @@ export default class SegmentLoader extends videojs.EventTarget {
   updateTimingInfoEnd_(segmentInfo) {
     segmentInfo.timingInfo = segmentInfo.timingInfo || {};
     const useVideoTimingInfo =
-      this.loaderType_ === 'main' && this.startingMedia_.hasVideo;
+      this.loaderType_ === 'main' && this.startingMedia_ && this.startingMedia_.hasVideo;
```
Comment (gesinger) [inline src/segment-loader.js:2451]: Since the check for startingMedia happens earlier, I think we can change this back

```suggestion
      this.loaderType_ === 'main' && this.startingMedia_.hasVideo;
```

File: src/util/container-request.js (hunk 17 above)
Comment (gesinger) [reply src/util/container-request.js:39]: I'm probably missing something, but I guess what I'm asking is, if we're trying to detect the container of a video mp4 segment, and we need 4 bytes for that detection, but we got 3 bytes on progress, will this return no detected type and the playlist get blacklisted?

File: test/test-helpers.js [hunk 24]
```diff
@@ -372,31 +373,29 @@ export const standardXHRResponse = function(request, data) {
...
-  if (isTypedBuffer) {
-    response = data.buffer;
-  } else if (!data) {
-    // A placeholder is used for some old tests. This may be a good target to clean up in
-    // the future.
-    response = new Uint8Array(1024).buffer;
-  } else {
-    response = data;
+  if (isTypedArray(response)) {
+    // a string for container requests or a buffer for non-container requests
```
Comment (gesinger) [inline test/test-helpers.js:397]:
```suggestion
    // a string for partial requests or a buffer for non-partial requests
```

File: src/util/container-request.js [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    beforeSend(request) {
+      // this forces the browser to pass the bytes to us unprocessed
+      request.overrideMimeType('text/plain; charset=x-user-defined');
+      request.addEventListener('progress', function(event) {
+        return callbackWrapper(request, null, {statusCode: request.status}, progressListener);
+      });
+    }
+  };
+
+  return xhr(options, () => {});
```
Comment (gesinger) [inline src/util/container-request.js:55]: Will this prevent normal (non progress) XHR completions? (If so, we may have to look into IE11, or worry about other browser compatibilities for progress: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestEventTarget/onprogress#Browser_compatibility .)

File: src/util/container-request.js (hunk 17 above)
Comment (brandonocasey) [reply src/util/container-request.js:39]: I think you may have a point her, going to refactor this a bit.

File: src/util/container-request.js (hunk 25 above)
Comment (brandonocasey) [reply src/util/container-request.js]: ie 11 on windows 8 and 10 does support progress events, but I updated the code anyway to run the final callback if the progress requests did not finish the request.

**Comment by brandonocasey** [comment]
I think I have all the comments handled here at one point, but I realized that we want another helper in vhs-utils, so we will have to get that in before we can finish this pr up. https://github.com/videojs/vhs-utils/pull/13

File: src/util/container-request.js [hunk 26]
```diff
@@ -1,58 +1,85 @@
...
+  return;
+};
+
 const containerRequest = (uri, xhr, cb) => {
-  let bytes;
+  let bytes = [];
   let id3Offset;
+  let finished = false;
+
+  const callback = function(err, req, type, _bytes) {
```
Comment (gesinger) [inline src/util/container-request.js:19]: Minor, but I find it a bit hard to read when there are two params that are essentially the same name, but have different spellings (e.g., `cb` and `callback`). Since this one serves the purpose of finishing the request, maybe we can name it something along the lines of `endRequestAndCallback`.

Comment (gkatsev) [reply src/util/container-request.js:19]: `requestCompletedCallback` here
and on line 14 `cb` to `containerRequestCallback`

Comment (brandonocasey) [reply src/util/container-request.js]: Done!
